tcp/src/dispatch.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright 2015, 2016 Ethcore (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use std;
+use std::net::SocketAddr;
+use std::sync::{Arc, Mutex};
+
+use futures::{task, Stream, Poll, Async};
+
+use std::collections::{HashMap, VecDeque};
+use std::collections::hash_map::Entry;
+
+
+pub type MessageQueue = Mutex<HashMap<SocketAddr, VecDeque<String>>>;
+pub type TaskNotificationQueue = Mutex<Vec<task::Task>>;
+
+pub struct PeerMessageQueue<S: Stream> {
```

Comment:
This stream produces same data as `up` stream + also adds messages from the queue

It parks current task when there are neither request responses nor queued messages and unparks once any message is pushed (task can also be unparked by the `up` stream itself - chacked in `message` test)